### PR TITLE
Con2d & MaxPool2d backward

### PR DESCRIPTION
**Pull request by vb000**
- Conv2d backward operator.
- MaxPool2d backward operator.
- Added `bsg_assert_msg` to print error messages with assert.
- Place read-only data section in dram, not dmem.

### CODE REVIEW

File: hammerblade/torch/hb_pytorch_link.ld [hunk 1]
```diff
@@ -0,0 +1,184 @@
+/******************************/
```
Comment (yodada) [inline hammerblade/torch/hb_pytorch_link.ld:1]: This looks very important. Could you maybe lecture us what is difference between dram and dmem?

File: hammerblade/torch/kernel/bsg_assert.hpp [hunk 2]
```diff
@@ -5,8 +5,13 @@
 #ifndef _BSG_ASSERT_HPP
 #define _BSG_ASSERT_HPP
 
-#define bsg_assert(cond) if (!(cond)) {                                          \
-    bsg_printf("assert failed at %s:%d\n", __FILE__, __LINE__);                \
+#define bsg_assert(cond) if (!(cond)) {                       \
+    bsg_printf("assert failed at %s:%d", __FILE__, __LINE__); \
+    bsg_fail();}
+
+#define bsg_assert_msg(cond, fmt, ...) if (!(cond)) {          \
```
Comment (yodada) [inline hammerblade/torch/kernel/bsg_assert.hpp:12]: sweet!

File: hammerblade/torch/tests/hbutils.py [hunk 3]
```diff
@@ -0,0 +1,6 @@
+def init_hb_tensor(input_t):
+    """
+    Returns a HB tensor that's a leaf node of the computation graph.
+    """
+    return input_t.hammerblade().clone().detach().requires_grad_(
```
Comment (yodada) [inline hammerblade/torch/tests/hbutils.py:5]: This is very smart! apparently I don't know much about using pytorch and ML in general. ;)

**Review by yodada — APPROVED** [review]
Looks good to me!

File: hammerblade/torch/hb_pytorch_link.ld (hunk 1 above)
Comment (vb000) [reply hammerblade/torch/hb_pytorch_link.ld:1]: `dmem` is the scratch pad and `dram` is the main memory of HB. Debug messages like printing and constants are stored in `rodata` section, which by default goes into `dmem` in Manycore's link script. We would want those to go to `dram` as scratchpad is a scarce resource.

Comment (yodada) [reply hammerblade/torch/hb_pytorch_link.ld:1]: ahh I see.

**Pull request merged**